### ISSUE TITLE
exception from `Uri.encodeComponent`

### ISSUE THREAD

**Issue by devoncarew** [dart-lang/dev_compiler#413]
I see an exception from `Uri.encodeComponent`. Some sample code:

```dart
Uri.encodeComponent('foo bar');
```

and the exception:

```
Uncaught Uncaught Error: CastError: Casting value of type function Uint8Array() { [native code] } to incompatible type NativeUint8List
Stack Trace:
Error
    at Object.throw_ [as throw] (web/ddc/dev_compiler/runtime/dart/_operations.js:308:23)
    at Object.throwCastError (web/ddc/dev_compiler/runtime/dart/_errors.js:24:21)
    at Object.cast [as as] (web/ddc/dev_compiler/runtime/dart/_operations.js:224:42)
    at Function._create1 (web/ddc/dev_compiler/runtime/dart/_native_typed_data.js:1171:19)
    at Function.new (web/ddc/dev_compiler/runtime/dart/_native_typed_data.js:1145:30)
    at Function.new (web/ddc/dev_compiler/runtime/dart/typed_data.js:69:52)
    at Function._createBuffer (web/ddc/dev_compiler/runtime/dart/convert.js:2087:38)
    at _Utf8Encoder.withBufferSize (web/ddc/dev_compiler/runtime/dart/convert.js:2082:36)
    at _Utf8Encoder.ctor (web/ddc/dev_compiler/runtime/dart/_classes.js:301:47)
    at Utf8Encoder.convert (web/ddc/dev_compiler/runtime/dart/convert.js:2048:21)
    at Utf8Codec.encode (web/ddc/dev_compiler/runtime/dart/convert.js:18:29)
    at Function._uriEncode (web/ddc/dev_compiler/runtime/dart/core.js:2970:28)
    at Function.encodeComponent (web/ddc/dev_compiler/runtime/dart/core.js:2817:18)
```

**Comment by jmesserly** [comment]
TL;DR -- typed_data isn't implemented yet.

We're probably trying to cast a "real" JS type to a fake interceptor type: https://github.com/dart-lang/dev_compiler/blob/f3c9328c6d0b9d5816eb3bcecddc7e218452859d/tool/input_sdk/private/native_typed_data.dart#L1266

(Our compiler doesn't/shouldn't understand the dart2js `@Native` syntax.)

A quick fix that might work is putting the correct DDC annotations on these types `@JSPeerInterface`, but I'd have to double check, I think we may still have some hardcoded assumptions about which types are mapped to native ones.

**Comment by devoncarew** [comment]
Ah, good to know. This isn't blocking me - I wrote around the call to `Uri.encodeComponent`.

**Comment by devoncarew** [comment]
To update, I hit one more place where I was calling Uri.encodeComponent. From talking with @vsmenon, typed_data's `Uint8Array` is probably the most widely used one (or at least, adding support for that specific typed_data class would address several issues I'm seeing).

**Comment by vsmenon** [comment]
The following now runs correctly for me as of #441:

```dart
void main() {
  var result = Uri.encodeComponent('foo bar');
  print(result);
}
```
